File: R/lollipopChart.R


#' Create a lollipop chart
#'
#' @param data The data frame containing the variables to consider.
#' @param x The categorical variable to consider. Will be plotted on the x-axis.
#' @param y The numeric variable to consider. Will be plotted on the y-axis.
#' @param sort Whether to sort or not the vertical lines. Takes three values
#' 'none' which is the default, 'ascending' or 'descending'.
#' @param bgcol The background-color of the SVG output. Defaults to 'white'.
#' @param xticks Optional. the number of x-axis ticks to consider.
#' @param yticks Optional. The number of y-axis ticks to consider.
#' @param xFontSize the font size of the x-axis labels. Defaults to 10.
#' @param yFontSize the font size of the y-axis labels. Defaults to 10.
#' @param font The font family to consider for the titles. Defaults to
#' "Verdana, Geneva, Tahoma, sans-serif".
#' @param xtitle Optional. The title of the x-axis.
#' @param xtitleFontSize The font size of the x-axis title. Defaults to 16.
#' @param ytitle Optional. The title of the y-axis.
#' @param ytitleFontSize The font size of the y-axis title. Defaults to 16.
#' @param title Optional. The title of the plot.
#' @param titleFontSize The font size of the plot title. Defaults to 22.
#' @param lineStroke The stroke color of the vertical lines. Defaults to 'maroon'.
#' @param lineStrokeWidth The vertical lines stroke's width. Defaults to 4.
#' @param circleFill The color of the circles. Defaults to 'lime'.
#' @param circleStroke The color of the stroke surrounding the circle. Defaults to 'lime'.
#' @param circleStrokeWidth The width of the circles' stroke. Defaults to 1.
#' @param circleRadius The radius of the circles. Defaults to 10.
#' @param axisCol the color of the x and y axis. It includes the ticks, the labels and titles.
#' Defaults to 'black'.
#' @param width Optional. The width of the SVG output.
#' @param height Optional. The height of the SVG output.

#' @return A SVG lollipop chart.
#' @export
#'
#' @examples
#' library(ggplot2) # needed for the mpg data frame
#' library(dplyr) # needed for data wrangling
#'
#' mpg %>% group_by(drv) %>%
#'   summarise(median_cty = median(cty)) %>%
#'   lollipopChart(
#'     x = "drv",
#'     y = "median_cty",
#'     sort = "ascending",
#'     xtitle = "drv variable",
#'     ytitle = "median cty",
#'     title = "Median cty per drv"
#'   )

lollipopChart <- function(data,
                         x,
                         y,
                         sort = 'none',
                         bgcol = "white",
                         xticks = NULL,
                         yticks = NULL,
                         xFontSize = 12,
                         yFontSize = 12,
                         font = "Verdana, Geneva, Tahoma, sans-serif",
                         xtitle = NULL,
                         xtitleFontSize = 14,
                         ytitle = NULL,
                         ytitleFontSize = 14,
                         title = NULL,
                         titleFontSize = 20,
                         lineStroke = "maroon",
                         lineStrokeWidth = 4,
                         circleFill = "lime",
                         circleStroke = "lime",
                         circleStrokeWidth = 1,
                         circleRadius = 10,
                         axisCol = "black",
                         width = NULL,
                         height = NULL

                         ) {



  if (is.null(data[[x]]) || is.null(data[[y]])) {
    stop("Please check that x and y belong to the specified data frame")
  }

  if (!(any(c('none',
              'ascending',
              'descending') %in% sort))) {
    stop("
    the sort parameter can only take the following values:
    'none', 'ascending' or 'descending'
       ")
  }

  if (grepl(";", font)) {
    stop("please remove the ';' character from your font argument")
  }

  r2d3::r2d3(
    data = data,
    script = system.file(
      "d3/scatterplot/lollipopchart.js",
      package = "ddplot"
    ),
    options = list(
      x = x,
      y = y,
      sort = sort,
      bgcol = bgcol,
      xticks = xticks,
      yticks = yticks,
      xFontSize = xFontSize,
      yFontSize = yFontSize,
      font = font,
      xtitle = xtitle,
      xtitleFontSize = xtitleFontSize,
      ytitle = ytitle,
      ytitleFontSize = ytitleFontSize,
      title = title,
      titleFontSize = titleFontSize,
      lineStroke = lineStroke,
      lineStrokeWidth = lineStrokeWidth,
      circleStroke = circleStroke,
      circleStrokeWidth = circleStrokeWidth,
      circleFill = circleFill,
      circleRadius = circleRadius,
      axisCol = axisCol,
      width = width,
      height = height
    )
  )

}

File: inst/d3/scatterplot/lollipopchart.js
// Setting the margin object
let margin = { top: 50, right: 50, bottom: 50, left: 50 };

// Sorting the data

data.sort(function (a, b) {
    if (options.sort === "none") {
      return null;
    } else if (options.sort === "ascending") {
      return d3.ascending(a[options.y], b[options.y]);
    } else {
      return d3.descending(a[options.y], b[options.y]);
    }
  })

// Setting the scaling of the x variable
let xScale = d3
  .scaleBand()
  .domain(
    data.map(function (d) {
      return d[options.x];
    })
  )
  .range([margin.left, width - margin.right])
  .padding(1);

// Setting the scaling of the y variable
let yScale = d3
  .scaleLinear()
  .domain([0, d3.max(data, (d) => d[options.y])])
  .range([height - margin.bottom, margin.top]);

// Setting the x-axis
xAx = d3.axisBottom(xScale).ticks(options.xticks);

// Setting the y-axis
yAx = d3.axisLeft(yScale).ticks(options.yticks);

// Rendering the x-axis
gx = svg
  .append("g")
  .attr("transform", "translate(0," + (height - margin.bottom) + ")")
  .call(xAx)
  .attr("font-size", options.xFontSize);

// Rendering the y-axis
gy = svg
  .append("g")
  .attr("transform", "translate(" + margin.left + ", 0)")
  .call(yAx)
  .attr("font-size", options.yFontSize);

// Removing the y and x axis line
gx.call((g) => g.selectAll("line").remove());
gy.call((g) => g.select(".domain").remove());

// Rendering the x-axis title
svg
  .append("text")
  .attr("transform", "translate(" + width / 2 + " ," + (height - 1) + ")")
  .attr("dx", "0em")
  .attr("dy", "-0.4em")
  .style("text-anchor", "middle")
  .style("font-family", options.font)
  .style("font-size", options.xtitleFontSize)
  .text(options.xtitle)
  .style('fill', options.axisCol)

// Rendering the y-axis title
svg
  .append("text")
  .attr("transform", "translate(" + 0 + " ," + height / 2 + ") rotate(-90)")
  .attr("dy", "1em")
  .style("text-anchor", "middle")
  .style("font-size", options.ytitleFontSize)
  .style("font-family", options.font)
  .text(options.ytitle)
  .style('fill', options.axisCol)

// Rendering the chart title
svg
  .append("text")
  .attr("x", width / 2)
  .attr("y", margin.top / 2)
  .attr("text-anchor", "middle")
  .attr("dx", "0em")
  .style("font-size", options.titleFontSize)
  .style("font-family", options.font)
  .text(options.title)
  .style('fill', options.axisCol)

// Modifying the color of the ticks and labels
svg.selectAll(".tick line").attr("stroke", options.axisCol)
svg.selectAll(".tick text").style("fill", options.axisCol)
svg.selectAll("path.domain").attr("stroke", options.axisCol)
  

// Rendering the lines

svg.style("background-color", options.bgcol);

svg
  .selectAll("myline")
  .data(data)
  .enter()
  .append("line")
  .attr("x1", (d) => xScale(d[options.x]))
  .attr("y1", (d) => yScale(d[options.y]))
  .attr("x2", (d) => xScale(d[options.x]))
  .attr("y2", yScale(0))
  .attr("stroke", options.lineStroke)
  .attr("stroke-width", options.lineStrokeWidth);

// Rendering the circles
svg
  .selectAll("mycircle")
  .data(data)
  .enter()
  .append("circle")
  .attr("cx", (d) => xScale(d[options.x]))
  .attr("cy", (d) => yScale(d[options.y]))
  .attr("r", options.circleRadius)
  .style("fill", options.circleFill)
  .attr("stroke", options.circleStroke)
  .attr("stroke-width", options.circleStrokeWidth);
